### DISCUSSION TITLE
[BUG] - Set-PnPListItem not writing all values for Multiple User fields

### DISCUSSION THREAD

**Discussion by simoncumming** [pnp/powershell#3754] [General]
### Reporting an Issue or Missing Feature
When running set-pnplistitem and trying to set the value for a User type field to 3 users, it will add only users #1 and #3. If trying to add 2 users, it will add only user #1

### Expected behavior 
When adding 3 values for a user field, I expected all 3 values to be populated for the item.

### Actual behavior 
When adding 3 values, I see values #1 and #3 added to the item, but not value #2

### Steps to reproduce behavior
Behaviour can be reproduced with a script similar to this:
`set-pnplistitem -List $ListName -Identity $id -Values @{"DocumentAuthor" = "AdeleV@v5wpc.onmicrosoft.com","GradyA@v5wpc.onmicrosoft.com","diegos@v5wpc.onmicrosoft.com"}
$item = Get-PnPListItem -List $ListName -Id $id
$item.FieldValues.DocumentAuthor`

The syntax of this is directly from: https://pnp.github.io/powershell/cmdlets/Set-PnPListItem.html#-values

### What is the version of the Cmdlet module you are running?
running version 2.3.0 

### Which operating system/environment are you running PnP PowerShell on?
- [X ] Windows

**Comment by gautamdsheth** [comment]
hi @simoncumming , can't reproduce the issue.

Is it possible that one of the users that you specified in the array is not added to User Site Information list ? If that's the case, then you need to "add" them to the site using `New-PnPUser` cmdlet

**Comment by simoncumming** [comment]
Hi @gautamdsheth can confirm that I've added all the users via New-PnPUser (should have included that in the sample script sorry).

I've also verified this by swapping the order:

`set-pnplistitem -List $ListName -Identity $id -Values @{"DocumentAuthor" = "AdeleV@v5wpc.onmicrosoft.com","GradyA@v5wpc.onmicrosoft.com","diegos@v5wpc.onmicrosoft.com"}      
$item = Get-PnPListItem -List $ListName -Id $id
$item.FieldValues.DocumentAuthor
`
RETURNS:

` Email       : AdeleV@v5wpc.onmicrosoft.com
TypeId      : {c956ab54-16bd-4c18-89d2-996f57282a6f}
LookupId    : 11
LookupValue : Adele Vance


Email       : DiegoS@v5wpc.onmicrosoft.com
TypeId      : {c956ab54-16bd-4c18-89d2-996f57282a6f}
LookupId    : 13
LookupValue : Diego Siciliani `


With the user values order changed:

`set-pnplistitem -List $ListName -Identity $id -Values @{"DocumentAuthor" = "AdeleV@v5wpc.onmicrosoft.com","diegos@v5wpc.onmicrosoft.com","GradyA@v5wpc.onmicrosoft.com"}      
$item = Get-PnPListItem -List $ListName -Id $id
$item.FieldValues.DocumentAuthor`

RETURNS:

` Email       : AdeleV@v5wpc.onmicrosoft.com
TypeId      : {c956ab54-16bd-4c18-89d2-996f57282a6f}
LookupId    : 11
LookupValue : Adele Vance


Email       : GradyA@v5wpc.onmicrosoft.com
TypeId      : {c956ab54-16bd-4c18-89d2-996f57282a6f}
LookupId    : 12
LookupValue : Grady Archie `


I am also having this issue across multiple tenants.

The prompt for me to check this was I found this forum post from someone in 2020 that was experiencing a similar issue [https://techcommunity.microsoft.com/t5/sharepoint-developer/add-pnplistitem-and-set-pnplistitem-multiple-user-field-adding/m-p/1962678/highlight/true#M9770](https://techcommunity.microsoft.com/t5/sharepoint-developer/add-pnplistitem-and-set-pnplistitem-multiple-user-field-adding/m-p/1962678/highlight/true#M9770)

I searched for an issue for it but couldn't find that they'd ever logged one.

**Comment by gautamdsheth** [comment]
Nope, still can't repro:

<img width="942" alt="image" src="https://github.com/pnp/powershell/assets/8832018/4e0c32dd-76eb-4d51-a851-fbbfa87a85bf">

and 

<img width="797" alt="image" src="https://github.com/pnp/powershell/assets/8832018/9f23fe9d-b9ef-496f-a2ec-6fde71455900">


@veronicageek - can you maybe try reproducing this issue ?

**Comment by veronicageek** [comment]
@gautamdsheth @simoncumming 

I cannot reproduce either (using 2.3.0)

![image](https://github.com/pnp/powershell/assets/25181757/ca621144-7599-446b-8600-221be4ff6c8e)

**Comment by simoncumming** [comment]
@gautamdsheth @veronicageek thanks for both looking into this, do you have any clues what could potentially cause something like this? Are there any settings or dependencies or something I should be looking at? Just trying to figure out the next troubleshooting steps

**Comment by gautamdsheth** [comment]
Am not sure about this. But looks like it could be an issue with users itself. Not sure how we can help here, am moving this to a discussion , maybe somebody in the community has some idea.

Don't see this as bug with the module itself.

**Comment by st1234b** [comment]
I just had the same problem.
The cause was that it was an array of type Object.
After I cast it to the type String Array, it worked.

PS D:\PowershellSkripte\Code> $siteData.Administratoren.GetType()
IsPublic IsSerial Name                                     BaseType
True     True     Object[]                                 System.Array

PS D:\PowershellSkripte\Code> [string[]]$sa = $siteData.Administratoren
PS D:\PowershellSkripte\Code> $sa.GetType()
IsPublic IsSerial Name                                     BaseType
True     True     String[]                                 System.Array

Set-PnPListItem -List "SharePointSiteAdmins" -Identity $spListItem.ID -Values @{"Title" = $siteData.Title; "SiteID" = $siteData.SiteID; "SiteAdmins" = $sa; "SiteDescription" = $siteData.Description }  -ErrorAction Stop

**Comment by SupraCupra** [comment]
I've run into this same issue with a client. 
This  had seemed to work before, and for a range of documents in this specific document library. However for this field no longer (i havent tested the others. 
I'm new to github, so bear with me.  

I've tried @st1234b's solution but still doesnt work 

[string[]]$adminUsers = "Name1@XX.com","Name2@XX.com"
$metadata["Admin_x002F_s"] = $adminUsers
Set-PnPListItem -List $targetLibrary -Identity $id -Values $metadata -UpdateType SystemUpdate

Just isolating on this one field seems to update the field with the first name in the array. Not sure why it isn't updating two. The script runs successfully with no errors but I can see the value isnt updated.

Anyone else have this issue?

Just updated my module and tried again with the same result
PnP.Powershell ver 3.1.0

UPDATE:


[string[]]$adminUsers = "Name1@XX.com","Name2@XX.com","Name3@XX.com"

Results in an update

The column now has "Name1@XX.com","Name3@XX.com" 


This is thanks to https://techcommunity.microsoft.com/discussions/sharepointdev/add-pnplistitem-and-set-pnplistitem-multiple-user-field---adding-more-than-one-s/1961918
